README.md
# Logstash Plugin

[![Travis Build Status](https://travis-ci.com/logstash-plugins/logstash-filter-cipher.svg)](https://travis-ci.com/logstash-plugins/logstash-filter-cipher)

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-filter-awesome.gemspec
```
- Install the plugin from the Logstash home
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.

CHANGELOG.md
## 4.0.3
  - Fix: Debug statement doesn't use .value fetch key's value [#33](https://github.com/logstash-plugins/logstash-filter-cipher/pull/33)

## 4.0.2
  - [DOC] Fixes "Note" formatting for the Key setting [#30](https://github.com/logstash-plugins/logstash-filter-cipher/pull/30)

## 4.0.1
  - General improvements to code and docs [#29](https://github.com/logstash-plugins/logstash-filter-cipher/pull/29)
    - Fixed threadsafety; this plugin can now be used in pipelines with more than one worker.
    - Fixed a potential leak of the configured key into logs; the key is now only included if trace-level logging is enabled.
    - Fixed an issue where configurations that used invalid `mode` or `algorithm` settings could produce unhelpful error messages.
    - Fixed an issue where a bad payload could cause the plugin to crash; when exceptions are encountered, the offending event will now be tagged with `_cipherfiltererror`.
    - Improved documentation substantially.

## 4.0.0
  - Removed obsolete iv field

## 3.0.1
  - Update gemspec summary

## 3.0.0
  - Mark deprecated iv field obsolete

## 2.0.7
  - Fix some documentation issues

## 2.0.5
 - internal,deps: Depend on logstash-core-plugin-api instead of logstash-core, removing the need to mass update plugins on major releases of logstash

## 2.0.4
 - internal,deps: New dependency requirements for logstash-core for the 5.0 release

## 2.0.3
 - bugfix: fixes base64 encoding issue, adds support for random IVs 

## 2.0.0
 - internal: Plugins were updated to follow the new shutdown semantic, this mainly allows Logstash to instruct input plugins to terminate gracefully, 
   instead of using Thread.raise on the plugins' threads. Ref: https://github.com/elastic/logstash/pull/3895
 - internal,deps: Dependency on logstash-core update to 2.0


spec/filters/cipher_spec.rb
# encoding: utf-8
require 'logstash-core'
require 'logstash/json'
require "logstash/codecs/base"
require 'logstash/filters/cipher'

describe LogStash::Filters::Cipher do

  let(:cleartext) do
    'شسيبشن٤٤ت٥ت داھدساققبمر фывапролдзщшгнекутцйячсмить asdfghjklqpoiuztreyxcvbnm,.-öäü+ä123ß´yö.,;LÖÜ*O 來少精清皆人址法田手扌打表氵開日大木裝 1234567890#$%^&*()!@#;:\'?.>,<testAESEn+=_-~`}]'
  end

  describe 'single event, encrypt/decrypt aes-128-cbc, 16b RANDOM IV, 16b key, b64 encode' do
    let(:event) do
      LogStash::Event.new(LogStash::Json.load("{\"message\":\"#{cleartext}\"}"))
    end

    let(:encrypter) do
      described_class.new(
        "algorithm" => "aes-128-cbc",
        "cipher_padding" => 1,
        "iv_random_length" => 16,
        "key" => "1234567890123456",
        "key_size" => 16,
        "mode" => "encrypt",
        "source" => "message",
        "target" => "message_crypted",
        "base64" => true,
        "max_cipher_reuse" => 1)
    end

    let(:decrypter) do
      described_class.new(
        "algorithm" => "aes-128-cbc",
        "cipher_padding" => 1,
        "iv_random_length" => 16,
        "key" => "1234567890123456",
        "key_size" => 16,
        "mode" => "decrypt",
        "source" => "message_crypted",
        "target" => "message_decrypted",
        "base64" => true,
        "max_cipher_reuse" => 1)
    end

    before(:each) do
      encrypter.register
      decrypter.register
    end

    let(:result) do
      encrypter.filter(event)
      decrypter.filter(event)
      event
    end

    it 'validate initial cleartext message' do
      expect(result.get("message")).to eq(cleartext)
    end

    it 'validate decrypted message' do
      expect(result.get("message_decrypted")).to eq(result.get("message"))
    end

    it 'validate encrypted message is not equal to message' do
      expect(result.get("message")).not_to eq(result.get("message_crypted"))
    end

  end

  describe '1000 events, 11 re-use, encrypt/decrypt aes-128-cbc, 16b RANDOM IV, 16b key, b64 encode' do

    total_events = 1000

    let(:events) do
      (1..total_events).map do |i|
        LogStash::Event.new(LogStash::Json.load("{\"message\":\"#{cleartext}\"}"))
      end
    end

    let(:encrypter) do
      described_class.new(
        "algorithm" => "aes-128-cbc",
        "cipher_padding" => 1,
        "iv_random_length" => 16,
        "key" => "1234567890123456",
        "key_size" => 16,
        "mode" => "encrypt",
        "source" => "message",
        "target" => "message_crypted",
        "base64" => true,
        "max_cipher_reuse" => 11)
    end

    let(:decrypter) do
      described_class.new(
        "algorithm" => "aes-128-cbc",
        "cipher_padding" => 1,
        "iv_random_length" => 16,
        "key" => "1234567890123456",
        "key_size" => 16,
        "mode" => "decrypt",
        "source" => "message_crypted",
        "target" => "message_decrypted",
        "base64" => true,
        "max_cipher_reuse" => 11)
    end

    before(:each) do
      encrypter.register
      decrypter.register
    end

    let(:results) do
      events.each do |e|
        encrypter.filter(e)
        decrypter.filter(e)
      end
      events
    end

    it 'validate total events' do
      expect(results.length).to eq(total_events)
    end

    it 'validate initial cleartext message' do
      results.each do |result|
        expect(result.get("message")).to eq(cleartext)
      end
    end

    it 'validate decrypted message' do
      results.each do |result|
        expect(result.get("message_decrypted")).to eq(result.get("message"))
      end
    end

    it 'validate encrypted message is not equal to message' do
      results.each do |result|
        expect(result.get("message")).not_to eq(result.get("message_crypted"))
      end
    end
  end
end


lib/logstash/filters/cipher.rb
# encoding: utf-8
require "logstash/filters/base"
require "openssl"
require "concurrent/atomic/thread_local_var"

# This filter parses a source and apply a cipher or decipher before
# storing it in the target.
#
class LogStash::Filters::Cipher < LogStash::Filters::Base
  config_name "cipher"

  # The field to perform filter
  #
  # Example, to use the @message field (default) :
  # [source,ruby]
  #     filter { cipher { source => "message" } }
  config :source, :validate => :string, :default => "message"

  # The name of the container to put the result
  #
  # Example, to place the result into crypt :
  # [source,ruby]
  #     filter { cipher { target => "crypt" } }
  config :target, :validate => :string, :default => "message"

  # Do we have to perform a `base64` decode or encode?
  #
  # If we are decrypting, `base64` decode will be done before.
  # If we are encrypting, `base64` will be done after.
  #
  config :base64, :validate => :boolean, :default => true

  # The key to use
  #
  # NOTE: If you encounter an error message at runtime containing the following:
  #
  # "java.security.InvalidKeyException: Illegal key size: possibly you need to install 
  # Java Cryptography Extension (JCE) Unlimited Strength Jurisdiction Policy Files for your JRE"
  #
  # Please read the following: https://github.com/jruby/jruby/wiki/UnlimitedStrengthCrypto
  #
  config :key, :validate => :password

  # The key size to pad
  #
  # It depends of the cipher algorithm. If your key doesn't need
  # padding, don't set this parameter
  #
  # Example, for AES-128, we must have 16 char long key. AES-256 = 32 chars 
  # [source,ruby]
  #     filter { cipher { key_size => 16 }
  #
  config :key_size, :validate => :number, :default => 16

  # The character used to pad the key
  config :key_pad, :default => "\0"

  # The cipher algorithm
  #
  # A list of supported algorithms can be obtained by
  # [source,ruby]
  #     puts OpenSSL::Cipher.ciphers
  config :algorithm, :validate => OpenSSL::Cipher.ciphers, :required => true

  # Encrypting or decrypting some data
  #
  # Valid values are encrypt or decrypt
  config :mode, :validate => %w(encrypt decrypt), :required => true

  # Cipher padding to use. Enables or disables padding.
  #
  # By default encryption operations are padded using standard block padding
  # and the padding is checked and removed when decrypting. If the pad
  # parameter is zero then no padding is performed, the total amount of data
  # encrypted or decrypted must then be a multiple of the block size or an
  # error will occur.
  #
  # See EVP_CIPHER_CTX_set_padding for further information.
  #
  # We are using Openssl jRuby which uses default padding to PKCS5Padding
  # If you want to change it, set this parameter. If you want to disable
  # it, Set this parameter to 0
  # [source,ruby]
  #     filter { cipher { cipher_padding => 0 }}
  config :cipher_padding, :validate => :string

  # Force an random IV to be used per encryption invocation and specify
  # the length of the random IV that will be generated via:
  #
  #       OpenSSL::Random.random_bytes(int_length)
  #
  # Enabling this will force the plugin to generate a unique
  # random IV for each encryption call. This random IV will be prepended to the
  # encrypted result bytes and then base64 encoded. On decryption "iv_random_length" must
  # also be set to utilize this feature. Random IV's are better than statically
  # hardcoded IVs
  #
  # For AES algorithms you can set this to a 16
  # [source,ruby]
  #     filter { cipher { iv_random_length => 16 }}
  config :iv_random_length, :validate => :number, :required => true

  # If this is set the internal Cipher instance will be
  # re-used up to @max_cipher_reuse times before being
  # reset() and re-created from scratch. This is an option
  # for efficiency where lots of data is being encrypted
  # and decrypted using this filter. This lets the filter
  # avoid creating new Cipher instances over and over
  # for each encrypt/decrypt operation.
  #
  # This is optional, the default is no re-use of the Cipher
  # instance and max_cipher_reuse = 1 by default
  # [source,ruby]
  #     filter { cipher { max_cipher_reuse => 1000 }}
  config :max_cipher_reuse, :validate => :number, :default => 1

  def register
    require 'base64' if @base64
    if cipher_reuse_enabled?
      @reusable_cipher = Concurrent::ThreadLocalVar.new
      @cipher_reuse_count = Concurrent::ThreadLocalVar.new
    end

    if @key.value.length != @key_size
      @logger.debug("key length is " + @key.value.length.to_s + ", padding it to " + @key_size.to_s + " with '" + @key_pad.to_s + "'")
      @key = @key.class.new(@key.value[0,@key_size].ljust(@key_size,@key_pad))
    end
  end # def register

  def filter(event)
    source = event.get(@source)
    if (source.nil? || source.empty?)
      @logger.debug("Event to filter, event 'source' field: " + @source + " was null(nil) or blank, doing nothing")
      return
    end

    result = case(@mode)
             when "encrypt" then do_encrypt(source)
             when "decrypt" then do_decrypt(source)
             else
               @logger.error("Invalid cipher mode. Valid values are \"encrypt\" or \"decrypt\"", :mode => @mode)
               raise "Internal Error, aborting."
             end

    event.set(@target, result)
    filter_matched(event) unless result.nil?
  rescue => e
    @logger.error("An error occurred while #{@mode}ing.", :exception => e.message)
    event.tag("_cipherfiltererror")
  end

  private

  def cipher_reuse_enabled?
    @max_cipher_reuse > 1
  end

  ##
  # @param plaintext [String]
  # @return [String]: ciphertext
  def do_encrypt(plaintext)
    with_cipher do |cipher|
      random_iv = OpenSSL::Random.random_bytes(@iv_random_length)
      cipher.iv = random_iv

      ciphertext = random_iv + cipher.update(plaintext) + cipher.final

      ciphertext = Base64.strict_encode64(ciphertext).encode("utf-8") if @base64 == true

      ciphertext
    end
  end

  ##
  # @param ciphertext_with_iv [String]
  # @return [String] plaintext
  def do_decrypt(ciphertext_with_iv)
    ciphertext_with_iv = Base64.strict_decode64(ciphertext_with_iv) if @base64 == true
    encoded_iv = ciphertext_with_iv.byteslice(0..@iv_random_length)
    ciphertext = ciphertext_with_iv.byteslice(@iv_random_length..-1)

    with_cipher do |cipher|
      cipher.iv = encoded_iv
      plaintext = cipher.update(ciphertext) + cipher.final
      plaintext.force_encoding("UTF-8")
      plaintext
    end
  end

  ##
  # Returns a new or freshly-reset cipher, bypassing cipher reuse if it is not enabled
  #
  # @yieldparam [OpenSSL::Cipher]
  # @yieldreturn [Object]: the object that this method should return
  # @return [Object]: the object that was returned by the yielded block
  def with_cipher
    return yield(init_cipher) unless cipher_reuse_enabled?

    with_reusable_cipher do |reusable_cipher|
      yield reusable_cipher
    end
  end

  ##
  # Returns a new or freshly-reset cipher.
  #
  # @yieldparam [OpenSSL::Cipher]
  # @yieldreturn [Object]: the object that this method should return
  # @return [Object]: the object that was returned by the yielded block
  def with_reusable_cipher
    cipher = get_or_init_reusable_cipher

    result = yield(cipher)

    cleanup_reusable_cipher

    return result
  rescue => e
    # when an error is encountered, we cannot trust the state of the cipher object.
    @logger.debug("shared cipher: removing because an exception was raised in #{Thread.current}", :exception => e.message)
    destroy_reusable_cipher
    raise
  end

  def get_or_init_reusable_cipher
    if @reusable_cipher.value.nil?
      @logger.debug("shared cipher: initializing for #{Thread.current}")
      @reusable_cipher.value = init_cipher
      @cipher_reuse_count.value = 0
    end

    @cipher_reuse_count.value += 1
    @reusable_cipher.value
  end

  def cleanup_reusable_cipher
    if @cipher_reuse_count.value >= @max_cipher_reuse
      @logger.debug("shared cipher: max_cipher_reuse[#{@max_cipher_reuse}] reached for #{Thread.current}, total_cipher_uses = #{@cipher_reuse_count.value}") if @logger.debug?
      destroy_reusable_cipher
    else
      @logger.debug("shared cipher: resetting for #{Thread.current}")
      @reusable_cipher.value.reset
    end
  end

  def destroy_reusable_cipher
    @reusable_cipher.value = nil
    @cipher_reuse_count.value = 0
  end

  ##
  # @return [OpenSSL::Cipher]
  def init_cipher
    cipher = OpenSSL::Cipher.new(@algorithm)

    cipher.public_send(@mode)

    cipher.key = @key.value

    cipher.padding = @cipher_padding if @cipher_padding

    if @logger.trace?
      @logger.trace("Cipher initialisation done", :mode => @mode,
                                                  :key => @key.value,
                                                  :iv_random_length => @iv_random_length,
                                                  :iv_random => @iv_random,
                                                  :cipher_padding => @cipher_padding)
    end

    cipher
  end # def init_cipher


end # class LogStash::Filters::Cipher
